### ISSUE TITLE
Tutorial model problems

### ISSUE THREAD

**Issue by agrigoriev** [remicres/otbtf#23]
Hi! First of all - thank you for your work. I'm trying to dive into the topic and your toolkit is very time-saving.

I'm trying to accomplish the tutorial.
When I use
https://gitlab.irstea.fr/remi.cresson/otbtf/-/blob/master/python/create_savedmodel_simple_cnn.py
of a model, I get an error during evaluation.

```
2020-06-02 18:28:24.866539: I tensorflow/cc/saved_model/loader.cc:333] SavedModel load for tags { serve }; Status: success: OK. Took 11351714 microseconds.
2020-06-02 18:28:24 (INFO) TensorflowModelTrain: New source:
2020-06-02 18:28:24 (INFO) TensorflowModelTrain: Patch size               : [16, 16]
2020-06-02 18:28:24 (INFO) TensorflowModelTrain: Placeholder (training)   : x
2020-06-02 18:28:24 (INFO) TensorflowModelTrain: Placeholder (validation) : x
2020-06-02 18:28:24 (INFO) TensorflowModelTrain: New source:
2020-06-02 18:28:24 (INFO) TensorflowModelTrain: Patch size               : [1, 1]
2020-06-02 18:28:24 (INFO) TensorflowModelTrain: Placeholder (training)   : y
2020-06-02 18:28:24 (INFO) TensorflowModelTrain: Tensor name (validation) : prediction
2020-06-02 18:28:24 (INFO) TensorflowModelTrain: Set validation mode to classification validation
Training epoch #1: 0% [                                                  ]2020-06-02 18:28:25.911561: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2020-06-02 18:28:26.619262: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
Training epoch #1: 100% [**************************************************] (5s)
Training epoch #2: 100% [**************************************************] (1s)
Training epoch #3: 100% [**************************************************] (1s)
Training epoch #4: 100% [**************************************************] (1s)
Training epoch #5: 100% [**************************************************] (1s)
Training epoch #6: 100% [**************************************************] (1s)
Training epoch #7: 100% [**************************************************] (1s)
Training epoch #8: 100% [**************************************************] (1s)
Training epoch #9: 100% [**************************************************] (1s)
Training epoch #10: 100% [**************************************************] (1s)
Evaluate model (Learning data): 100% [**************************************************]2020-06-02 18:28:43 (FATAL) TensorflowModelTrain: itk::ERROR: Number of elements in the tensor is 4 but image outputRegion has 100 values to fill.
Buffer region:
ImageRegion (0x7ffe900ba830)
  Dimension: 2
  Index: [0, 0]
  Size: [1, 100]

Number of components: 1
Tensor shape:
 {4}
Please check the input(s) field of view (FOV), the output field of expression (FOE), and the  output spacing scale if you run the model in fully convolutional mode (how many strides in your model?)
```

With this model:
https://github.com/remicres/otbtf_tutorials_resources/blob/master/01_patch_based_classification/models/create_model1.py
Which is very close in terms of model parameters, everything works fine.

And one more question - Do you have any tutorial for FCN case? 
And in case if I have dense annotations - polygons in separate shp - what is the correct pipeline for data preparation?

**Comment by remicres** [comment]
It might be a bug related to the (new) Keras API of TF. I am investigating!

**Comment by remicres** [comment]
It should be corrected now. Can you give a try?
If you use the doker image, you can update the code like this:

```
cd /work/otb/otb/Modules/Remote/otbtf
git pull
```

**Comment by agrigoriev** [comment]
It works now.

**Comment by remicres** [comment]
Great. 
Regarding your last question, some tutorials will come soon!

**Comment by agrigoriev** [comment]
One more question.

With the new model, all metrics are equal to 1, though training and test
data is different.


2020-06-03 12:18:29 (INFO) TensorflowModelTrain: Metrics for target
"prediction":
2020-06-03 12:18:29 (INFO) TensorflowModelTrain: Precision of class [0] vs
all: 1
2020-06-03 12:18:29 (INFO) TensorflowModelTrain: Recall of class [0] vs
all: 1
2020-06-03 12:18:29 (INFO) TensorflowModelTrain: F-score of class [0] vs
all: 1
2020-06-03 12:18:29 (INFO) TensorflowModelTrain:
2020-06-03 12:18:29 (INFO) TensorflowModelTrain: Precision of the different
classes: [1]
2020-06-03 12:18:29 (INFO) TensorflowModelTrain: Recall of the different
classes: [1]
2020-06-03 12:18:29 (INFO) TensorflowModelTrain: F-score of the different
classes: [1]
2020-06-03 12:18:29 (INFO) TensorflowModelTrain:
2020-06-03 12:18:29 (INFO) TensorflowModelTrain: Kappa index: 1
2020-06-03 12:18:29 (INFO) TensorflowModelTrain: Overall accuracy index: 1
2020-06-03 12:18:29 (INFO) TensorflowModelTrain: Confusion matrix:
39753


Best regards,
Grigoriev Alexey

**Comment by remicres** [comment]
Strange: it looks like there only a single class in your dataset. Are you following [this tutorial](https://mdl4eo.irstea.fr/2019/01/04/an-introduction-to-deep-learning-on-remote-sensing-images-tutorial/)? There should be 6 classes.

**Comment by agrigoriev** [comment]
Is there a way to prepare labels for FCN networks, like, say, U-net?
I have the shp file and need to prepare both paches and labels with ground
truth.

Best regards,
Grigoriev Alexey

**Comment by remicres** [comment]
Of course it is possible. Rasterize your shp over your image using the OTB `Rasterization` application for instance. Then set `OTB_TF_NSOURCE` to `2`, extract your patches using `PatchesExtraction`, and follow the same logic as in the tutorial. Then you can use your model in fully convolutional mode. However please open a dedicated issue for this matter, or just wait a few days/weeks that I put a toturial online :wink:
